gearcmd/worker.go

```diff
@@ -28,59 +30,80 @@ type TaskConfig struct {
 	RetryCount   int
 }
 
+var (
+	lg = logger.New("gearcmd")
+)
+
 // Process runs the Gearman job by running the configured task.
 // We need to implement the Task interface so we return (byte[], error)
 // though the byte[] is always nil.
-func (conf TaskConfig) Process(job baseworker.Job) ([]byte, error) {
-	// This wraps the actual processing to do some logging
-	log.Println(kayvee.FormatLog("gearcmd", kayvee.Info, "START",
-		map[string]interface{}{"function": conf.FunctionName, "job_id": getJobID(job), "job_data": string(job.Data())}))
+func (conf TaskConfig) Process(job baseworker.Job) (b []byte, returnErr error) {
+	jobID := getJobID(job)
+	if jobID == "" {
+		jobID = strconv.Itoa(rand.Int())
+		lg.InfoD("rand-job-id", logger.M{"msg": "no job id parsed, random assigned."})
+	}
+
+	data := logger.M{
+		"function": conf.FunctionName,
+		"job_id":   jobID,
+		"job_data": string(job.Data()),
+	}
 
+	// This wraps the actual processing to do some logging
+	lg.InfoD("START", data)
 	start := time.Now()
-	for {
-		err := conf.doProcess(job)
-		end := time.Now()
-		data := map[string]interface{}{
-			"function": conf.FunctionName,
-			"job_id":   getJobID(job),
-			"job_data": string(job.Data()),
-			"type":     "gauge",
+
+	for try := 0; try < conf.RetryCount+1; try++ {
+		// We create a temporary directory to be used as the work directory of the process.
+		// A new work directory is created for every retry of the process.
+		// We try to use MEOS_SANDBOX, the default will be the system temp directory.
+		tempDirPath, err := ioutil.TempDir(os.Getenv("MESOS_SANDBOX"),
+			fmt.Sprintf("%s-%s-%d-", conf.FunctionName, jobID, try))
+		if err != nil {
+			lg.CriticalD("tempdir-failure", logger.M{"error": err.Error()})
+			return nil, err
 		}
+		defer os.RemoveAll(tempDirPath)
```

Comment:
this won't clean up directories from failures, right? should we do that explicitly when we retry or should we put the whole retry loop in a function?